kbase-extension/static/kbase/js/api/RestAPIClient.js

```diff
@@ -58,63 +58,73 @@ define(['jquery'], ($) => {
         // gawd. What i wouldn't give for some ES6 syntax here.
         routeKeys.forEach((routeName) => {
             const route = args.routes[routeName];
-
             const routeArgs = route.path.match(/\${\w+}/g) || [];
 
-            if (routeArgs) {
-                for (let j = 0; j < routeArgs.length; j++) {
-                    const remapped = routeArgs[j].match(/\w+/g)[0];
-                    routeArgs[j] = remapped;
-                }
+            for (let j = 0; j < routeArgs.length; j++) {
+                const remapped = routeArgs[j].match(/\w+/g)[0];
+                routeArgs[j] = remapped;
             }
 
             this[routeName] = function (fArgs) {
                 if (fArgs === undefined) {
                     fArgs = {};
                 }
 
+                const ajaxArgs = Object.assign(
+                    {},
+                    {
+                        token: this.token,
+                        route,
+                    },
+                    fArgs
+                );
+
                 let path = route.path;
+                for (const element of routeArgs) {
+                    const replacement = fArgs[element] !== undefined ? fArgs[element] : '';
+                    path = path.replace('${' + element + '}', replacement);
+                }
 
-                for (let i = 0; i < routeArgs.length; i++) {
-                    const replacement =
-                        fArgs[routeArgs[i]] !== undefined ? fArgs[routeArgs[i]] : '';
-                    path = path.replace('${' + routeArgs[i] + '}', replacement);
+                // add in the Auth header
+                if (!ajaxArgs.headers) {
+                    ajaxArgs.headers = {};
                 }
-                const restURL = [this.root, path].join('/');
-
-                return this.ajax({
-                    route: route,
-                    url: restURL,
-                    token: this.token,
-                    data: fArgs.data,
-                });
+                ajaxArgs.headers.Authorization = this.token;
+
+                ajaxArgs.url = [this.root, path].join('/');
+                return this.ajax(ajaxArgs);
             };
         });
 
-        this.ajax = function ajax(args) {
+        /**
+         * AJAX query executed by each route created above
+         *
+         * @param {object} ajaxArgs arguments to pass to $.ajax, including:
+         *    dataType  {string}    sets the 'Accept' header
+         *    data      {any}
+         *    headers   {object}    object containing header names and values
+         *                          e.g. {'Content-Type': 'application/json', ...}
+         * @returns jQuery promise
+         */
+        this.ajax = function ajax(ajaxArgs) {
             const deferred = $.Deferred();
+            const promise = deferred.promise();
 
-            const beforeSend = function (xhr) {
-                xhr.setRequestHeader('Authorization', args.token);
-            };
```

Comment:
not sure why this was done using `beforeSend`, since you can easily do it with the `headers` argument instead